### PR TITLE
Blockingly shut-down webrender api

### LINKED ISSUE #26876: Frequent webrender panics when shutting down
I often get backtraces that look like this:
```
called `Option::unwrap()` on a `None` value (thread WRSceneBuilder#0, at /Users/jdm/.cargo/git/checkouts/webrender-c3596abe1cf4f320/de39995/webrender/src/scene_builder_thread.rs:616)
   0: backtrace::backtrace::trace_unsynchronized
   1: <servo::backtrace::Print as core::fmt::Debug>::fmt
   2: core::fmt::write
   3: std::io::Write::write_fmt
   4: servo::backtrace::print
   5: servo::main::{{closure}}
   6: std::panicking::rust_panic_with_hook
   7: _rust_begin_unwind
   8: core::panicking::panic_fmt
             at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libcore/panicking.rs:89
   9: core::panicking::panic
             at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libcore/panicking.rs:52
  10: core::option::Option<T>::unwrap
  11: webrender::scene_builder_thread::SceneBuilderThread::process_transaction
  12: webrender::scene_builder_thread::SceneBuilderThread::run::{{closure}}
  13: core::iter::adapters::map_fold::{{closure}}
  14: core::iter::traits::iterator::Iterator::fold::ok::{{closure}}
  15: core::iter::traits::iterator::Iterator::try_fold
  16: core::iter::traits::iterator::Iterator::fold
  17: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
  18: core::iter::traits::iterator::Iterator::for_each
  19: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::spec_extend
  20: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
  21: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
  22: core::iter::traits::iterator::Iterator::collect
  23: webrender::scene_builder_thread::SceneBuilderThread::run
  24: webrender::renderer::Renderer::new::{{closure}}
  25: std::sys_common::backtrace::__rust_begin_short_backtrace
  26: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
  27: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  28: std::panicking::try::do_call
  29: ___rust_try
  30: std::panicking::try
  31: std::panic::catch_unwind
  32: std::thread::Builder::spawn_unchecked::{{closure}}
  33: core::ops::function::FnOnce::call_once{{vtable.shim}}
  34: std::sys::unix::thread::Thread::new::thread_start
  35: __pthread_start
```
I suspect this is because we end up sending webrender transactions to webrender after we've officially started shutting down, but the IPC router thread doesn't know about that so it continues to forward them.

### PR DESCRIPTION
**Pull request by gterzian**
FIX https://github.com/servo/servo/issues/26876

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by jdm** [comment]
```
1532 unexpected results that are NOT known-intermittents:
  ▶ CRASH [expected PASS] /css/CSS2/box-display/anonymous-boxes-inheritance-001.xht
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  │ Shutting down the Constellation after generating an output file or exit flag specified
  │ called `Result::unwrap()` on an `Err` value: "SendError(..)" (thread main, at /root/.cargo/git/checkouts/webrender-c3596abe1cf4f320/1175aca/webrender_api/src/api.rs:1558)
  │    0: servo::backtrace::print
  │    1: servo::main::{{closure}}
  │    2: std::panicking::rust_panic_with_hook
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/panicking.rs:481
  │    3: rust_begin_unwind
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/panicking.rs:385
  │    4: core::panicking::panic_fmt
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libcore/panicking.rs:89
  │    5: core::option::expect_none_failed
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libcore/option.rs:1272
  │    6: webrender_api::api::RenderApi::shut_down
  │    7: compositing::compositor::IOCompositor<Window>::handle_browser_message
  │    8: compositing::compositor::IOCompositor<Window>::receive_messages
  │    9: servo::Servo<Window>::handle_events
  │   10: servo::app::App::handle_events
  │   11: servo::app::App::run
  │   12: servo::main
  │   13: std::rt::lang_start::{{closure}}
  │   14: std::rt::lang_start_internal::{{closure}}
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/rt.rs:52
  │       std::panicking::try::do_call
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/panicking.rs:297
  │       std::panicking::try
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/panicking.rs:274
  │       std::panic::catch_unwind
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/panic.rs:394
  │       std::rt::lang_start_internal
  │              at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/libstd/rt.rs:51
  │   15: main
  │   16: __libc_start_main
  │   17: _start
  └ [2020-06-15T15:02:22Z ERROR servo] called `Result::unwrap()` on an `Err` value: "SendError(..)"
```
Every failure looks like this.

**Comment by gterzian** [comment]
Ok I think it's because the compositor and the webgl thread are using the same webrender(webgl only has a sender initially, however it is used to get an api via `webrender_api_sender.create_api()`, and those "api" are just handles to the same backend it seems), so trying to only shut-down in the webgl thread.

**Comment by jdm** [comment]
The previous try successfully completed: https://treeherder.allizom.org/#/jobs?repo=servo-try&revision=3ed66ad00b3434c47812c1a145c8e114c89782d3

File: components/canvas/webgl_thread.rs [hunk 1]
```diff
@@ -333,6 +333,8 @@ impl WebGLThread {
                         Ok(msg) => {
                             let exit = self.handle_msg(msg, &webgl_chan);
                             if exit {
+                                // Block on shutting-down Webrender.
```
Comment (atouchet) [inline components/canvas/webgl_thread.rs:336]: Really nitpicky but the first R in WebRender is capitalized.

**Comment by gterzian** [comment]
@jdm r? 

I haven't tried reproducing the problem locally, however I am fairly certain this will fix it, and if not we can re-open the issue?

**Comment by jdm** [comment]
```
  ▶ CRASH [expected PASS] /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ 
  │ 
  │ true
  │ Shutting down the Constellation after generating an output file or exit flag specified
  │ called `Result::unwrap()` on an `Err` value: "SendError(..)" (thread WebGL thread, at /Users/worker/.cargo/git/checkouts/webrender-c3596abe1cf4f320/1175aca/webrender_api/src/api.rs:1653)
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::write
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::main::{{closure}}
  │    5: std::panicking::rust_panic_with_hook
  │    6: _rust_begin_unwind
  │    7: core::panicking::panic_fmt
  │    8: core::result::unwrap_failed
  │    9: webrender_api::api::RenderApi::send_transaction
  │   10: canvas::webgl_thread::WebGLThread::remove_webgl_context
  │   11: <canvas::webgl_thread::WebGLThread as core::ops::drop::Drop>::drop
  │   12: core::ptr::drop_in_place
  │   13: std::sys_common::backtrace::__rust_begin_short_backtrace
  │   14: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   15: std::sys::unix::thread::Thread::new::thread_start
  │   16: __pthread_body
  │   17: __pthread_start
```

The Drop implementation for WebGLThread conflicts with this change. We should probably ensure that we remove contexts before shutting down webrender if we're explicitly shutting down.

Then again, if we're already destroying the WebGLThread (either because of shutting down or panicking), does it matter if we clean up resources at all?

**Comment by gterzian** [comment]
Thanks, I moved that logic to the explicit exit branch(we can remove it later if it is confirmed as unnecessary). 

I'm still a little worried the compositor might still use the WR API after it has been shut-down by canvas, although I guess we can address it if/when it happens(the current situation is a panic on shutdown as well).

**Comment by gterzian** [comment]
```
3 unexpected results that are NOT known-intermittents:
  ▶ CRASH [expected TIMEOUT] /_webgl/conformance2/renderbuffers/readbuffer.html
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ [2020-06-16T12:37:51Z ERROR canvas::webgl_thread] Last GL operation failed: ReadBuffer(1029)
  │ Unexpected WebGL error: 0x502 (1282) [ReadBuffer(1029)] (thread WebGL thread, at components/canvas/webgl_thread.rs:2241)
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::write
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::main::{{closure}}
  │    5: std::panicking::rust_panic_with_hook
  │    6: _rust_begin_unwind
  │    7: std::panicking::begin_panic_fmt
  │    8: canvas::webgl_thread::WebGLImpl::apply
  │    9: canvas::webgl_thread::WebGLThread::handle_msg
  │   10: canvas::webgl_thread::WebGLThread::process
  │   11: std::sys_common::backtrace::__rust_begin_short_backtrace
  │   12: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   13: std::sys::unix::thread::Thread::new::thread_start
  │   14: __pthread_body
  │   15: __pthread_start
  │ [2020-06-16T12:37:51Z ERROR servo] Unexpected WebGL error: 0x502 (1282) [ReadBuffer(1029)]
  │ Should have destroyed the FBO renderbuffers with `destroy()`! (thread WebGL thread, at /Users/worker/.cargo/git/checkouts/surfman-d15b9f3e86c2e263/2283aca/surfman/src/renderbuffers.rs:25)
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::write
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::main::{{closure}}
  │    5: std::panicking::rust_panic_with_hook
  │    6: std::panicking::begin_panic
  │    7: <surfman::renderbuffers::Renderbuffers as core::ops::drop::Drop>::drop
  │    8: core::ptr::drop_in_place
  │    9: core::ptr::drop_in_place

```

"Should have destroyed the FBO renderbuffers with `destroy()`!", happens at https://github.com/servo/surfman/blob/ae14cc9afd140bc73e89f3fe8c080f729f09ccd2/surfman/src/renderbuffers.rs#L25

My guess is this is related to surfman-chains? 

Perhaps the drop was never actually run, and now that I've moved the clean-up code to "earlier" on exit, the problem emerges? 

We do seem to destroy all swapchains, but maybe not all?

cc @asajeffrey

**Comment by gterzian** [comment]
Ah no I'm guessing the actual panic is "Unexpected WebGL error: 0x502 (1282) [ReadBuffer(1029)] (thread WebGL thread, at components/canvas/webgl_thread.rs:2241)", and then the buffers drop without having been destroyed...

**Comment by gterzian** [comment]
Ah so these are all `CRASH [expected TIMEOUT]`. Not sure how the change here would have unblocked a timeout however.

**Comment by jdm** [comment]
Right, only the first panic matters. It's not clear to me how the changes in this PR can trigger a webgl behaviour change, though.

**Comment by gterzian** [comment]
It's not clear to me either, however I can imagine that properly sutting-down WR might prevent some build-up of system resources during the test run, which previously caused those test to eventually start timing out.

**Comment by jdm** [comment]
You're right, I hadn't noticed that every test is TIMEOUT->CRASH. I'm pretty sure all of these tests were panicking already, so I don't think there's any actual regression in behaviour here. Go ahead and update the test results.

**Comment by gterzian** [comment]
@jdm r?

**Comment by gterzian** [comment]
Ah ok try=wpt is kind of useless in this case...

**Comment by gterzian** [comment]
That one looks like a new intermittent: `/fetch/cross-origin-resource-policy/image-loads.html`, it's passing locally on Mac, no issue seems to have been filed yet...

**Comment by CYBAI** [comment]
☝️ @gterzian I guess you might want this?

**Comment by CYBAI** [comment]
@bors-servo try- try=wpt-mac

- ah, no, I remembered a wrong command; trying wpt on mac should be `wpt-mac` 😂

**Comment by gterzian** [comment]
@jdm are you ok with https://github.com/servo/servo/issues/26944? 

I wouldn't say it's cause by this change, although apparently it hasn't been noted before either...

**Comment by jdm** [comment]
@bors-servo retry
* #25513

**Pull request merged**